### ISSUE TITLE
Transactions Failing When Utilizing Extension

### ISSUE THREAD

**Issue by arimgibson** [olivierwilkinson/prisma-extension-soft-delete#24]
When using the soft delete client to try and perform a transaction, the following error is thrown:

error: All elements of the array need to be Prisma Client promises. Hint: Please make sure you are not awaiting the Prisma client calls you intended to pass in the $transaction function.

The same transaction will run successfully when not utilizing the extension. **Workaround: utilize a normal Prisma client for transactions**

Can come back later this week to provide a reproduction; don't have time right this second and thought I'd throw this out there in case this is something we're already aware of. Couldn't find in other bug reports

**Comment by rpaterson** [comment]
The error appears on [sequential](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#sequential-prisma-client-operations) (array) style calls to `$transaction([ ... ])`, but not on [interactive](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#interactive-transactions) style calls to `$transaction(async (prisma) => { ... })`

**Comment by rpaterson** [comment]
So as a workaround it's easy to convert from "sequential" to "interactive" style

**Comment by rpaterson** [comment]
Actually looking at the code I think "interactive" style transactions don't throw this error, but the queries are not actually run inside the transaction, which is even worse!

**Comment by baojr** [comment]
I just found the same problem @rpaterson described in my project. After installing this extension, previously working code is now unable to read data written within the same transaction, which violates the isolation properties of SQL transactions.

**Comment by vinnymac** [comment]
I ran into this issue today as well. Although in my case, this only appears to happen if I configure additional extensions. If `prisma-extension-soft-delete` is the only extension configured, transactions work just fine 🤷🏼‍♂️.

**Comment by arimgibson** [comment]
> I ran into this issue today as well. Although in my case, this appears to also happen if I configure additional extensions. If `prisma-extension-soft-delete` is the only extension configured, transactions work just fine 🤷🏼‍♂️.
> 
> In my case, both removing `prisma-extension-soft-delete`, or making it the only extension resolves the issue.

Interesting observation; haven't seen the same here! This is the only one we're using and still having issues. Will try to get around to that reproducible example this weekend; keeps slipping my mind

For others with the same, please +1 this comment instead of starting a long chain of just "same"  -- keeps things cleaner. Thank you :pray:

**Comment by olivierwilkinson** [comment]
Hi all, thank you for raising this! Sorry for the slow response, I've been trying to figure out exactly what the problem is!

Unfortunately at this point I think this is a limitation in the Prisma extensions API. This extension changes the operation type from deletes to updates, and this requires using the previous instance of the client, the one that is being extended. When using interactive transactionn the client that needs to be used is the one that is provided in the $transaction callback, which is not the one used when changing operations due to this.

There was a separate but related issue which seemed to be fixed by replacing each client query method to use the previous client (#14), however this has broken the fluent API (#22).

My current plan is to try to create a repro in the Prisma repo to prove this and add some tests, however if I am right it might take a while to resolve by the Prisma team due to it requiring an API change to extensions.

For now I recommend using the [middleware version of this library](https://github.com/olivierwilkinson/prisma-soft-delete-middleware) which has feature parity with this one. I know that middleware is deprecated but I remember reading that the Prisma team are not intending to remove middleware in the foreseeable future, they deprecated them to encourage people to move to extensions. Due to middleware supporting changing operations I don't think these issues will be present, but I will add the tests provided in #25 to double check that. (Thank you @rpaterson for raising that PR!)

When using middleware with extensions the middleware needs to be applied to the initial (non-extended) client, however for the soft-delete behaviour in these libraries I don't think that will be a problem. There is a note about the [semantics of using middleware with extensions here](https://www.prisma.io/docs/orm/prisma-client/client-extensions/query#if-you-use-the-query-extension-component-and-middlewares).

If anybody has any thoughts or concerns surrounding this please feel free to reply to this! I really appreciate everyones patience with this issue as it is a major one! ❤️

**Comment by vinnymac** [comment]
I can confirm switching to the middleware works with transactions correctly. Even while using additional extensions.

Thanks for investigating it!

**Comment by BobReid** [comment]
I spent the day looking into this and will share my findings. 

There are 2 distinct issues going on with transactions, one of the transaction array API and the other on the interactive API. 

**Transaction Array API**

This extension seems to utalize the model method overrides as opposed to the query overrides. 
By that I mean the extension defines its overrides along the lines of 
```ts
prisma.$extends({
  model: {
    [Model]: { 
      count({...})  { ... }, 
        ... 
     ]
  }
}
```
as opposed to 
```ts 
prisma.$extends({
  query: {
    [Model]: { 
      count({...})  { ... }, 
        ... 
     ]
  }
}
```

When the model extensions are used the result is always a Promise as opposed to a [PrismaPromise](https://www.prisma.io/docs/orm/reference/prisma-client-reference#prismapromise-behavior). I tried many different techniques including returning an unresolved PrismaPromise and it always gets unwrapped and rewrapped as a vanilla JS Promise. 

PrismaPromises are the deferred promises that prisma uses to delay sql exection, allowing them to be collected and run in a transaction together, as opposed to firing off when they are first created. Vanilla Promises cannot be passed to the transaction API for this reason. 

**Interactive Transaction Issue**

@olivierwilkinson your intuition on the interaction transaction issue is close but not 100% correct. The operation switching from delete to update is a problem. But I don't believe it is because you are switching client instances (extended vs non-extended). 

The issue is that when the operation is switched it is called on a client directly instead of a transaction that the API requires. As far as I can tell there is no api to access the current transaction. 

Consider the code needed to create and then delete a user
```ts
prisma.$transaction((tx) => {
  const user = await tx.user.create(...)
  await tx.user.delete({where: { id: user.id } })
});  
```

When the operation is rewritten it is the equivalent of 
```ts
prisma.$transaction((tx) => {
  const user = await prisma.user.create(...)
  await prisma.user.delete({where: { id: user.id } })
});  
```

This code will execute but the statements are not attached to the transaction itself. 

Until the extensions API exposes the current transaction there is no way to safely rewrite the operation.

**Comment by olivierwilkinson** [comment]
Hi all,

I'm unfortunately back to thinking that it is some interaction between how different extensions are handling transactions. I'm going to make another concerted effort to fix this issue, I would really appreciate if everyone could list the extensions they are using so I can look into thier implementation and create a reproduction.

@BobReid thank you for looking into this! I think you are correct that my intuition was not quite right, unfortunately I am not using the extended client in the same way as in your snippet:
```typescript
const extended = prisma.$extends({
  query: {
    [Model]: {
      delete(...) { 
        return extended[Model].update(...)
      }
    }
  }
})
```
I am instead using the previous client using the `defineExtension` helper:
```typescript
Prisma.defineExtension((client) => {
  return client.$extends({
    query: {
      [Model]: {
        delete(...) {
          return client[Model].update(...)
        }
      }
    }
  });
});
```
When I tried to reproduce the problem in the Prisma repo using the second snippet everything seems to work fine so I think there is some deeper reason for the issue, possibly related to nested transactions: https://github.com/prisma/prisma/issues/19346

**Comment by rpaterson** [comment]
The failing test cases I created in #25 reproduce the problems without any other extensions.

**Comment by olivierwilkinson** [comment]
> The failing test cases I created in #25 reproduce the problems without any other extensions.

Ah yes you are right! Sorry! I'm getting a little ahead of myself and thinking about the issue with row level security which was also related to transactions (and might be reintroduced when I fix the fluent API).

Anyways that is not directly related to this issue and I will focus on fixing the tests in your PR, thank you again for doing that.

It might be that they are helpful producing a repro in the Prisma repo if that is where the issue lies, I was focused on rollback before but your tests are more nuanced.

**Comment by BobReid** [comment]
@olivierwilkinson 

Yes I know my snippet was different than your code. My snippet was trying to test using the extended client which still doesn't work. 

The issue is that neither the un-extended, not the extended client are the transaction instance that is passed as a argument to the interaction transaction block. 

IMO the extension API should pass the client used to make the call (either the based client or the transaction if the operation is invoked on a transaction) as an argment to the extension method. 

something like 

```ts
Prisma.defineExtension((client) => {
  return client.$extends({
    query: {
      [Model]: {
        delete(client, ..rest) {
          return client.update(...)
        }
      }
    }
  });
});
```

I briefly browsed the prisma source code, it seems trivial to do, but we would need to convince the team that this is the correct approach.

**Comment by vinnymac** [comment]
Prisma just removed middleware support in v6.14.0

https://www.prisma.io/blog/prisma-orm-v6-14-0-relationships-for-sql-views-more-robust-management-api-and-more

So this issue is now blocking my team, as we can’t rely on falling back to middleware as a fix any longer. I will be testing how extensions work in that version and report back.

**Comment by vinnymac** [comment]
Here is the error I get when using v2.0.1 of this extension with v6.14.0 of prisma:

```log
TypeError: Cannot read properties of undefined (reading 'forEach')
    at /Users/vinnymac/Sites/soft-delete-error-repro/node_modules/prisma-extension-soft-delete/dist/lib/helpers/createParams.js:11:24
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/vinnymac/Sites/soft-delete-error-repro/node_modules/prisma-extension-soft-delete/dist/lib/helpers/createParams.js:8:39)
    at Module._compile (node:internal/modules/cjs/loader:1688:14)
    at Object..js (node:internal/modules/cjs/loader:1820:10)
    at Module.load (node:internal/modules/cjs/loader:1423:32)
    at Function._load (node:internal/modules/cjs/loader:1246:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:235:24)
    at Module.<anonymous> (node:internal/modules/cjs/loader:1445:12)

Node.js v22.18.0
```

Which is from this line:
https://github.com/olivierwilkinson/prisma-extension-soft-delete/blob/9180d9854a58db9960bd6a491056fcf1c0e28c8b/src/lib/helpers/createParams.ts#L13

Looking at the model its working on I see `name`, `fields`, and `dbName`. So `uniqueFields` is now gone for some reason, maybe this object was moved or removed?

This issue is likely unrelated to the transaction problem specifically, but I am blocked by it in order to see whether or not we still get the issue reported by OP.

EDIT: For now I am working around the uniqueFields issue with this [patch here](https://gist.github.com/vinnymac/0fd3ee4c9ed5655fbffd9e48c6c9a1fd) so I can do further testing on this specific issue.

**Comment by vinnymac** [comment]
Okay, so the good news is after testing the [patch](https://gist.github.com/vinnymac/0fd3ee4c9ed5655fbffd9e48c6c9a1fd) I was able to use both `$transaction(async)` and `$transaction([])` apis without any errors being thrown with this extension.

So as long as we fix that `uniqueFields` bug I think we are going to be able to move from middleware to extension without any issues. Would be good for others to test this as well though, as I am sure people have different use cases I am not testing for.

Note: I also tested .delete() and .deleteMany() both of which worked for me still.

EDIT: it seems we may need #37 to fix this, possibly in conjunction with `await getDmmf()` from @prisma/internals. Since you would need the full dmmf to get access to the fields used by this extension. It’s a shame that Prisma broke popular extensions like this one without any regard for the community while simultaneously deprecating middleware in a minor version update.

**Comment by canastro** [comment]
I need to upgrade Prisma to resolve other issues, but I’m currently blocked because Prisma dropped middleware support and this library still depends on it to function correctly.

Is there any way we can help prioritize this work? If donations would help move this issue forward, I’d be happy to contribute.

**Comment by arimgibson** [comment]
Sorry for falling off the map and not responding here! Slipped my mind until today, when I was looking to implement this same extension in a different codebase and found this thread when doing some due diligence, lol

Agree so far with @BobReid 's interpretation, this is a limitation I've come across with Prisma before where the transaction client isn't available... something cool on Prisma's side would be storing the transaction client in AsyncLocalContext, and then reading from that when it exists, basically allowing all downstream operations (including this extension) to leverage the transaction client itself. But I digress, that would be complicated for them to implement.

@vinnymac it sounds like you might've uncovered a different issue, likely related to the `prisma-client` generator opposed to the `prisma-client-js` generator. Can you confirm the issue was caused not just by upgrading Prisma, but by switching to the new generator? Are you still having transaction issues apart from that?

Going to set aside some time to make a reproduction on the newer Prisma versions but using the `prisma-client-js` so we aren't running into the issue in #37

**Considering middleware is deprecated, I still think that using a version of the Prisma client without the extension would be the best approach**. This is what we did at my last company, which also allows you to maintain a client for hard deletes considering this extension doesn't have a method to hard delete currently. Code is something like this:

```typescript
export const nonSoftDeleteClient = new PrismaClient();

export const db = nonSoftDeleteClient.$extends(
  createSoftDeleteExtension({
    models: ...
  }),
);

// Usage
// Regular operations, soft deleting
await db.post.delete(...)

// Hard deletes outside of transaction
await nonSoftDeleteClient.post.delete(...)

// Transactions
await nonSoftDeleteClient.$transaction((tx) => {
  // Soft deleting without the extension
  await tx.post.update({
     where: { id: post.id },
     data: { deletedAt: new Date() }
  });

  // Hard delete, since the extension is not being used for this transaction
  await tx.post.delete();
})
```

**Comment by vinnymac** [comment]
Yes, that was with the `prisma-client` generator. I use both `prisma-client-js` and `prisma-client` depending on the project / use case.

I have since done exactly what you're describing and use a prisma client instance without this extension in most of my projects.
